### PR TITLE
funcgraph paint-up / paint-down / paint-merge (ie. stop where higher level branch remerges)

### PR DESCRIPTION
**Pull request by atlas0fd00m**
this branch:
* pulls functionality out of getLongPath that creates a histogram of weighted nodes and make it available to other code
* creates hot buttons and menu items in Function Graph to:
    * paint nodes up from the selected basic block to the beginning of the graph
    * paint nodes down from the selected basic block to the end of the graph
    * paint nodes down to the next remerge from other parts of the graph
* updates unittests to fix a few bugs, and make VIVBINS="path" use "path" as the base path (instead of test_vivisect/bins/).  
    * if you set VIVBINS=1  the old "test_vivisect/bins" is used instead.

### CODE REVIEW

File: vivisect/qt/funcgraph.py [hunk 1]
```diff
@@ -379,11 +394,12 @@ def _buttonSaveAs(self):
...
         file('test.html','wb').write(str(h))
 
-    def renderFunctionGraph(self, fva):
+    def renderFunctionGraph(self, fva, graph=None):
 
         self.fva = fva
         #self.graph = self.vw.getFunctionGraph(fva)
-        self.graph = viv_graphutil.buildFunctionGraph(self.vw, fva, revloop=True)
+        if graph == None:
+            self.graph = viv_graphutil.buildFunctionGraph(self.vw, fva, revloop=True)
```
Comment (invisig0th) [inline vivisect/qt/funcgraph.py]: Looks as though specifying graph!=None actually doesn't use it for anything...

Comment (atlas0fd00m) [inline vivisect/qt/funcgraph.py]: woops.  fixed.

**Comment by atlas0fd00m** [comment]
should be gtg.

**Comment by atlas0fd00m** [comment]
bump

File: vivisect/tools/graphutil.py [hunk 2]
```diff
@@ -15,26 +15,39 @@
 
 xrskip = envi.BR_PROC | envi.BR_DEREF
 
-def getLongPath(g, maxpath=1000):
-    '''
-    Returns a list of list tuples (node id, edge id) representing the longest path
+def getNodeWeightHisto(g):
```
Comment (atlas0fd00m) [inline vivisect/tools/graphutil.py]: this functionality is used in multiple places, and is valuable to have split out.

File: vivisect/tools/graphutil.py [hunk 3]
```diff
@@ -612,6 +625,33 @@ def getGraphNodeByVa(fgraph, va):
             return nva
     return None
 
+def findRemergeDown(graph, va):
```
Comment (atlas0fd00m) [inline vivisect/tools/graphutil.py]: track through a graph to where it remerges after a split.  very useful for compartmentalizing code chunks.

File: vivisect/tools/graphutil.py [hunk 4]
```diff
@@ -622,35 +662,38 @@ def preRouteGraph(graph, fromva, tova):
     preRouteGraphUp(graph, tova)
     preRouteGraphDown(graph, fromva)
 
-def clearGraphRouting(graph):
```
Comment (atlas0fd00m) [inline vivisect/tools/graphutil.py]: allow for clearing of specific tags, change graph routing to apply specific tags for up and down, and handle looping

File: vivisect/tools/graphutil.py [hunk 5]
```diff
@@ -658,16 +701,47 @@ def preRouteGraphDown(graph, fromva):
...
+        for eid, fr, to, einfo in graph.getRefsFrom(curnode):
+            if graph.getNodeProps(to).get(mark) == True:
                 continue
-            #raw_input("try next")
+            if not loop and nwlist.get(fr) >= nwlist.get(to):
+                continue
+
             todo.append(graph.getNode(to))
 
+def clearMarkDown(graph, fromva, loop=False, mark='up'):
```
Comment (atlas0fd00m) [inline vivisect/tools/graphutil.py]: very specific clearing of tags (properties).  very useful for marking subsets of graph-flow and then clearing a different subset of graph-flow, almost like a ven diagram but for graph-flow

File: vivisect/qt/funcgraph.py [hunk 6]
```diff
@@ -15,12 +15,15 @@
 
 from PyQt4.QtCore   import pyqtSignal, QPoint
 from PyQt4          import QtCore, QtGui, QtWebKit
-from vqt.main       import idlethread, idlethreadsync, eatevents, vqtconnect, workthread
```
Comment (atlas0fd00m) [inline vivisect/qt/funcgraph.py]: create the gui interface

File: vivisect/tests/testvivgraph.py [hunk 7]
```diff
@@ -3,23 +3,10 @@
 import vivisect
 import vivisect.tests.vivbins as vivbins
 import vivisect.tools.graphutil as viv_graph
+from vivisect.tests.vivbins import getTestWorkspace, getAnsWorkspace
 
 class VivGraphTest(unittest.TestCase):
 
-    def getTestWorkspace(self, fname, analyze=True):
```
Comment (atlas0fd00m) [inline vivisect/tests/testvivgraph.py]: these functions are important to multiple unittests

**Comment by atlas0fd00m** [comment]
bump

File: vivisect/tools/graphutil.py [hunk 8]
```diff
@@ -622,44 +662,38 @@ def preRouteGraph(graph, fromva, tova):
     preRouteGraphUp(graph, tova)
     preRouteGraphDown(graph, fromva)
 
-def clearGraphRouting(graph):
+def clearGraphRouting(graph, marks=['up','down']):
```
Comment (invisig0th) [inline vivisect/tools/graphutil.py]: no mutable types as default args plz

File: vivisect/tests/vivbins.py [hunk 9]
```diff
@@ -20,7 +25,12 @@ def getTestWorkspace(fname, analyze=True):
     return vw
 
 def getAnsWorkspace(fname):
-    fpath = os.path.join('test_vivisect','bins','%s.viv' % fname)
+    path = os.getenv('VIVBINS') 
+    fpath = os.path.join(path,"%s.viv" % fname)
+
+    if path == '1':
```
Comment (invisig0th) [inline vivisect/tests/vivbins.py]: on second thought... lets not... ;)

File: vivisect/tools/graphutil.py [hunk 10]
```diff
@@ -15,26 +15,39 @@
...
         if not len(g.getRefsFromByNid(cb)):
-            # todo is a tuple of (cb, current path, visited nodes)
+            # leaves is a tuple of (cb, current path, visited nodes)
             # these are our leaf nodes
-            todo[weight].append( (cb, list(), set()) ) 
-        # set code block to weight
-        cb_to_weights[ cb ] = weight 
+            leaves[weight].append( (cb, list(), set()) ) 
+
+        # create FFT
```
Comment (invisig0th) [inline vivisect/tools/graphutil.py]: newp

File: vivisect/tools/graphutil.py [hunk 11]
```diff
@@ -612,6 +625,33 @@ def getGraphNodeByVa(fgraph, va):
...
+def findRemergeDown(graph, va):
+    '''
+    starting at a given va, figure out the nodes connecting va to the next place something remerges
+    '''
+    startnid = getGraphNodeByVa(graph, va)
+
+    # paint down graph, 
+    preRouteGraphDown(graph, startnid, mark='hit', loop=False)
+
+    fft, nodewts, leaves = getNodeWeightHisto(graph)
```
Comment (invisig0th) [inline vivisect/tools/graphutil.py]: g/fft/r/?

File: vivisect/tools/graphutil.py (hunk 8 above)
Comment (invisig0th) [inline vivisect/tools/graphutil.py]: move to delNodesProps() on Graph base ( and use getNodesByProp )

**Comment by atlas0fd00m** [comment]
visi: please review this PR

**Comment by atlas0fd00m** [comment]
visi: i killed the specialized unit-testing code and moved it into it's own branch.  please mainline this branch.

File: visgraph/graphcore.py [hunk 12]
```diff
@@ -303,6 +303,14 @@ def delNodeProp(self, node, prop):
                 self.nodeprops[prop].pop(pval,None)
         return pval
 
+    def delNodesProps(graph, props=('up','down')):
+        '''
+        Delete all listed properties from all nodes in the graph.
+        '''
+        for prop in props:
+            for node in graph.getNodesByProp(prop):
+                graph.delNodeProp(node, mark)
```
Comment (invisig0th) [inline visgraph/graphcore.py]: no way this runs...  "mark" is undeclared here...

File: visgraph/graphcore.py [hunk 13]
```diff
@@ -303,6 +303,14 @@ def delNodeProp(self, node, prop):
                 self.nodeprops[prop].pop(pval,None)
         return pval
 
+    def delNodesProps(graph, props=('up','down')):
```
Comment (invisig0th) [inline visgraph/graphcore.py]: please make the self reference named correctly...

These seem like use-case specific defaults in what should be a general purpose API.  Should probably be:

```python
def delNodesProps(self, props):
```

**Comment by atlas0fd00m** [comment]
PR.

File: visgraph/graphcore.py [hunk 14]
```diff
@@ -303,6 +303,14 @@ def delNodeProp(self, node, prop):
                 self.nodeprops[prop].pop(pval,None)
         return pval
 
+    def delNodesProps(self, props=('up','down')):
```
Comment (invisig0th) [inline visgraph/graphcore.py]: These default args are super use case specific and probably exist as a hold-over from the function being within the other module.  props in this case should not be a kwarg.

**Pull request merged**